### ISSUE TITLE
Can't persist nested objects to Realm.

### ISSUE THREAD

**Issue by DeadPenguin** [realm/realm-dotnet#1866]
## Goals

Sorry for crossposting this from Stackoverflow, but I can't find anything in the documentation, nor could I find any solutions from similar problems. Also since this is important for my degree-project I'm kinda short on time. If this is too far off the intended use feel free to remove this.

I'm setting up a database for my Xamarin.Forms-project (going for UWP and Android) with some 1-to-many relationships but I just can't get Realm to persist the nested lists to the databse. All single-value-attributes are persisted perfectly, but not the collections.
I have already searched SO and other sites. I tried both advised versions of writing (Write-Function and Transaction) but to no avail.

I wish to add an instance of the book

## Code Sample

`public class Book : RealmObject
    {
        public Book()
        {
            BookTags = new List<string>();
            EntryTags = new HashSet<string>();
            Entries = new List<Entry>();
        }

        [PrimaryKey, Indexed]
        public int Id { get; set; }
        public int ParentId { get; set; }
        [Indexed]
        public string Title { get; set; }
        public string Author { get; set; }
        public IList<string> BookTags { get; }
        public ISet<string> EntryTags { get; }
        public IList<Entry> Entries { get; }
    }

    public class Entry : RealmObject
    {
        public Entry()
        {
            Tags = new List<Tag>();
        }

        [PrimaryKey, Indexed]
        public string Name { get; set; }
        [Backlink(nameof(Book.Entries))]
        public IQueryable<Book> Guide { get; }
        public string ImagePath { get; set; }
        public IList<Tag> Tags { get; }
    }

    public class Tag : RealmObject
    {
        [PrimaryKey, Indexed]
        public string Name { get; set; }
        public string Value { get; set; }
    }`

The most recent version of my AddBook-function which is supposed to write the book to the realm is as follows (I think you can ignore everything concerning the IDs but I wanted to include it anyway so you don't miss anything):
`public void AddBook(Book newBook, int parentId)
        {
            var temp = LocalRealm.All<IdManager>().SingleOrDefault();
            IdManager idManager = temp == null ? new IdManager() : (IdManager)temp;
            newBook.Id = idManager.NextId;
            newBook.ParentId = parentId;

            LocalRealm.Write(() =>
            {
                var book = new Book();
                book.Title = newBook.Title;
                book.Author = newBook.Author;
                book.Id = newBook.Id;
                book.ParentId = newBook.ParentId;

                foreach (string s in newBook.BookTags)
                    book.BookTags.Add(s);
                foreach (string s in newBook.EntryTags)
                    book.EntryTags.Add(s);
                foreach (Entry e in newBook.Entries)
                    book.Entries.Add(e);

                LocalRealm.Add(book);
                idManager.Inc();
                LocalRealm.Add(idManager, update: true);
            });
        }`

A previous version, where I simply set `var book = newBook` didn't work either. As I said: Id, ParentId, Title and Author are persisted just fine, while BookTags, EntryTags and Entries get lost during the Write-process. I already checked the values of book but they all seem to be set correctly inside the Write-block within both versions of the AddBook-function.

## Version of Realm and Tooling

- Realm Object Server Version: I'm using Realm version 3.4.0
- Flavor:
  - [ ] Realm Cloud
  - [ ] Self-Hosted
- Server OS & Version: -
- Client SDK Version: .NET 4.6.2
- Client OS & Version: Windows 10 v1809

**Comment by fealebenpae** [comment]
Don't initialize the collections in the constructors - you're clobbering the values Realm injects into the objects. Also, `ISet<T>` is not a supported Realm collection type - only `IList<T>` is.

Also, in your `AddBook` method why are you creating a new `Book` instance and copying the values from the `newBook` parameter into it before `Realm.Add`-ing it? Why not just `Realm.Add` `newBook` directly?

**Comment by DeadPenguin** [comment]
Thank you so much for the fast reply. The way I understood it, you have to create the object you want to persist inside the Write-Statement. The reason I initialized the Collections was, that VS throws an error, when I try to add a { set; } (saying setters are not supported yet) to any of the collections, but if I don't initialize them, I can't add any values outside of the write statement. 
Would it be better not to use the RealmObject-classes outside of Writes and create equivalent Structs for each of them? Up till now I used an instance of Book for two-way binding of all it's attributes to the UI, so the user can set them.
Also if you want to see my code in a better format, here is the corresponding [SO-Question](url).

**Comment by nirinchev** [comment]
Realm will automatically initialize the collections for you, don't worry. We plug into the msbuild pipeline (using [Fody](https://github.com/Fody/Fody)) to overwrite the getters of properties on RealmObject descendants. For example:

```csharp
class Person : RealmObject
{
    public IList<Dog> Dogs { get; }
}

// In your code
var person = new Person();

// When the object is unmanaged (not persisted in Realm)
// you can add objects/modify them outside of write transaction
person.Dogs.Add(new Dog());

realm.Write(() =>
{
    // Adding the object to the Realm needs to happen in a write transaction.
    realm.Add(person);
});

// When the object is managed, you can't modify it outside of write transaction.
// Adding objects to its list counts as modifying, so this will throw:
// person.Dogs.Add(new Dog());

// But there's no problem to add/remove objects inside another write transaction:
realm.Write(() =>
{
    person.Dogs.Add(new Dog());
});
```

Hope this clarifies it.

**Comment by DeadPenguin** [comment]
Thanks for your help! I resolved the issue by implementing your advices (i.e. removing Collection-Initializations and the {set;}-options from the Collections). For interaction with the UI (to get and later edit the values) I created equivalent, non-managed classes with the same attributes. Everytime I add something to Realm I convert it to an instance of the corresponding RealmObject-Class and add that. Might not be the most elegant solution but it seems to work fine.
